pgx/src/pgbox.rs

```diff
@@ -80,46 +82,62 @@ use std::ops::{Deref, DerefMut};
 ///     // we can't free it
 /// }
 /// ```
-///
-///
-/// ## Safety
-///
-/// TODO:
-///  - Interatctions with Rust's panic!() macro
-///  - Interactions with Postgres' error!() macro
-///  - Boxing a null pointer -- it works ::from_pg(), ::into_pg(), and ::to_pg(), but will panic!() on all other uses
-///
-pub struct PgBox<T> {
-    inner: Inner<T>,
+#[repr(transparent)]
+pub struct PgBox<T, AllocatedBy: WhoAllocated<T> = AllocatedByPostgres> {
+    ptr: Option<NonNull<T>>,
+    __marker: PhantomData<AllocatedBy>,
 }
 
-struct Inner<T> {
-    ptr: Option<*mut T>,
-    allocated_by_pg: bool,
+pub trait WhoAllocated<T> {
+    fn free(ptr: *mut T);
 }
+pub struct AllocatedByPostgres;
+pub struct AllocatedByRust;
 
-impl<T: Debug> Debug for PgBox<T> {
-    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), Error> {
-        match self.inner.ptr {
-            Some(ptr) => f.write_str(&format!(
-                "PgBox<{}> (ptr={:?}, owner={:?})",
-                std::any::type_name::<T>(),
-                unsafe {
-                    ptr.as_ref()
-                        .expect("impl Debug for PgBox expected self.ptr to be non-NULL")
-                },
-                self.owner_string()
-            )),
-            None => f.write_str(&format!(
-                "PgBox<{}> (ptr=NULL, owner={:?})",
-                std::any::type_name::<T>(),
-                self.owner_string()
-            )),
+impl<T> WhoAllocated<T> for AllocatedByPostgres {
+    /// Doesn't do anything
+    #[inline]
```

Comment:
Unnecessary, I think

Reply:
The impl itself?  No, it's necessary as the `Drop` impl does `AllocatedBy::free(...)`:  https://github.com/zombodb/pgx/blob/2b62ea2ee5905ebf1c638beccecda7dbc5ccc1a9/pgx/src/pgbox.rs#L332-L338

Reply:
I was referring to the inline